hedera-node/src/main/java/com/hedera/services/state/virtual/IterableStorageUtils.java

```diff
@@ -119,9 +140,54 @@ public static ContractKey upsertMapping(
 	public static @Nullable ContractKey removeMapping(
 			@NotNull final ContractKey key,
 			@NotNull final ContractKey root,
-			final VirtualMap<ContractKey, IterableContractValue> storage
+			@NotNull final VirtualMap<ContractKey, IterableContractValue> storage
 	) {
-		return MapValueListUtils.removeFromMapValueList(key, root, new ContractStorageListRemoval(key.getContractId(), storage));
+		return overwritingRemoveFromMapValueList(key, root, new ContractStorageListRemoval(key.getContractId(), storage));
 	}
 
+	private static ContractKey internalUpsertMapping(
+			@NotNull final ContractKey key,
+			@NotNull final IterableContractValue value,
+			@Nullable final ContractKey rootKey,
+			@Nullable final IterableContractValue rootValue,
+			@NotNull final VirtualMap<ContractKey, IterableContractValue> storage,
+			boolean useGetForModify
+	) {
+		final IterableContractValue oldValue;
+		if (useGetForModify) {
+			oldValue = storage.getForModify(key);
+		} else {
+			final var candidateValue = storage.get(key);
+			// Note it is ONLY safe to call copy() here---making the map's value
+			// immutable!---because we immediately put() the mutable value below
+			oldValue = (candidateValue != null) ? candidateValue.copy() : null;
+		}
+		if (oldValue != null) {
+			oldValue.setValue(value.getValue());
+			if (!useGetForModify) {
+				storage.put(key, oldValue);
+			}
+			return rootKey;
+		} else {
+			storage.put(key, value);
+			if (rootKey != null) {
+				value.setNextKey(rootKey.getKey());
+				final IterableContractValue nextValue;
+				if (rootValue != null) {
+					nextValue = rootValue;
+				} else {
+					nextValue = useGetForModify
+							? Objects.requireNonNull(storage.getForModify(rootKey), () -> "Missing root " + rootKey)
+							// Note it is ONLY safe to call copy() here---making the map's value
+							// immutable!---because we immediately put() the mutable value below
+							: Objects.requireNonNull(storage.get(rootKey), () -> "Missing root " + rootKey).copy();
+				}
+				nextValue.setPrevKey(key.getKey());
+				if (!useGetForModify && rootValue == null) {
```

Comment:
why do we need these two lines , if `rootValue == null` , we are getting `nextValue` from `storage.get(rootKey)` right ?